src/update/UpdateManager.cpp

```diff
@@ -156,16 +161,41 @@ void UpdateManager::GetUpdates(bool ignoreNetworkErrors)
                                     nullptr,    // parent
                                     wxPD_APP_MODAL | wxPD_ELAPSED_TIME | wxPD_SMOOTH));
 
+                                const wxPlatformInfo& info = wxPlatformInfo::Get();
+                                if (info.GetOperatingSystemId() & wxOS_WINDOWS)
+                                    audacityPatchFilename += ".exe";
+                                else if(info.GetOperatingSystemId() & wxOS_MAC)
+                                    audacityPatchFilename += ".dmg";
+
                                 auto audacityInstallerPath = wxFileName(
                                     wxStandardPaths::Get().GetUserDir(wxStandardPaths::Dir_Downloads)
                                     + FileNames::GetPathSeparator()
                                     + audacityPatchFilename);
+
+                                mAudacityInstaller.open(audacityInstallerPath.GetFullPath().ToStdString(), std::ios::binary);
                                 
                                 });
 
                             mProgressDialog->Pulse();
                         }
                     );
+
+                    downloadResponse->setOnDataReceivedCallback([downloadResponse, this](audacity::network_manager::IResponse*) {
+                        if (downloadResponse->getError() == audacity::network_manager::NetworkError::NoError)
+                        {
+                            std::vector<char> buffer(downloadResponse->getBytesAvailable());
+
+                            size_t bytes = downloadResponse->readData(buffer.data(), buffer.size());
+                            //NOTE need add error checkpoints: if (bytes != downloadResponse->getBytesAvailable())
+
+                            if (mAudacityInstaller.is_open())
+                            {
+                                mAudacityInstaller.write(buffer.data(), buffer.size());
```

Comment:
This block of code looks like the downloader first makes a huge memory buffer, tens of megabytes, to copy the download into completely, and then your code copies all of that into a file.

This looks inefficient to me:  can't the downloaded bytes from the internet go straight to the file instead?

But to fix that would need changes in the CurlResponse class.  Its interface doesn't support it.

I think it would need a "setWriteCallback" member function that it does not yet have.  But this would give you the finer control.

Discuss my idea with crsib.